File: stripe/src/main/java/com/stripe/android/model/Wallet.java
package com.stripe.android.model;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.json.JSONObject;

import static com.stripe.android.model.StripeJsonUtils.optString;

abstract class Wallet extends StripeJsonModel {
    private static final String FIELD_TYPE = "type";

    @Nullable
    static Wallet fromJson(@Nullable JSONObject walletJson) {
        if (walletJson == null) {
            return null;
        }

        final Type walletType = Type.fromCode(optString(walletJson, FIELD_TYPE));
        if (walletType == null) {
            return null;
        }

        return create(walletType, walletJson);
    }

    @Nullable
    private static Wallet create(@NonNull Type walletType, @NonNull JSONObject walletJson) {
        switch (walletType) {
            case AmexExpressCheckout: {
                // TODO(mshafrir): implemented
                return null;
            }
            case ApplePay: {
                // TODO(mshafrir): implemented
                return null;
            }
            case GooglePay: {
                // TODO(mshafrir): implemented
                return null;
            }
            case Masterpass: {
                // TODO(mshafrir): implemented
                return null;
            }
            case Microsoft: {
                // TODO(mshafrir): implemented
                return null;
            }
            case SamsungPay: {
                // TODO(mshafrir): implemented
                return null;
            }
            case VisaCheckout: {
                // TODO(mshafrir): implemented
                return null;
            }
            default: {
                return null;
            }
        }
    }

    enum Type {
        AmexExpressCheckout("amex_express_checkout"),
        ApplePay("apple_pay"),
        GooglePay("google_pay"),
        Masterpass("master_pass"),
        Microsoft("microsoft"),
        SamsungPay("samsung_pay"),
        VisaCheckout("visa_checkout");

        @NonNull private final String mCode;

        Type(@NonNull String code) {
            mCode = code;
        }

        @Nullable
        static Type fromCode(@Nullable String code) {
            for (Type type : values()) {
                if (type.mCode.equals(code)) {
                    return type;
                }
            }

            return null;
        }
    }
}

File: payments-core/src/main/java/com/stripe/android/model/wallets/Wallet.kt
package com.stripe.android.model.wallets

import android.os.Parcelable
import androidx.annotation.RestrictTo
import com.stripe.android.core.model.StripeModel
import com.stripe.android.model.Address
import kotlinx.parcelize.Parcelize

/**
 * Details of the card wallet.
 *
 * [card.wallet](https://stripe.com/docs/api/payment_methods/object#payment_method_object-card-wallet)
 */
sealed class Wallet(
    @get:RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
    val walletType: Type
) : StripeModel {

    @Parcelize
    data class AmexExpressCheckoutWallet
    @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
    constructor(
        val dynamicLast4: String?
    ) : Wallet(Type.AmexExpressCheckout)

    @Parcelize
    data class ApplePayWallet internal constructor(
        val dynamicLast4: String?
    ) : Wallet(Type.ApplePay)

    @Parcelize
    data class GooglePayWallet
    @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
    constructor(
        val dynamicLast4: String?
    ) : Wallet(Type.GooglePay), Parcelable

    @Parcelize
    data class MasterpassWallet internal constructor(
        val billingAddress: Address?,
        val email: String?,
        val name: String?,
        val shippingAddress: Address?
    ) : Wallet(Type.Masterpass)

    @Parcelize
    data class SamsungPayWallet internal constructor(
        val dynamicLast4: String?
    ) : Wallet(Type.SamsungPay)

    @Parcelize
    data class VisaCheckoutWallet internal constructor(
        val billingAddress: Address?,
        val email: String?,
        val name: String?,
        val shippingAddress: Address?,
        val dynamicLast4: String?
    ) : Wallet(Type.VisaCheckout)

    @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
    enum class Type(val code: String) {
        AmexExpressCheckout("amex_express_checkout"),
        ApplePay("apple_pay"),
        GooglePay("google_pay"),
        Masterpass("master_pass"),
        SamsungPay("samsung_pay"),
        VisaCheckout("visa_checkout");

        internal companion object {
            internal fun fromCode(code: String?): Type? {
                return values().firstOrNull { it.code == code }
            }
        }
    }
}
